crates/libs/windows/src/core/hstring.rs

```diff
@@ -420,27 +415,27 @@ impl Header {
         // TODO: allow this failure to propagate
         let header = heap_alloc(alloc_size).expect("Could not successfully allocate for HSTRING") as *mut Header;
 
+        // SAFETY: uses `std::ptr::write` (since `header` is unintialized). `Header` is safe to be all zeros.
         unsafe {
-            (*header).flags = 0;
+            header.write(std::mem::MaybeUninit::<Header>::zeroed().assume_init());
             (*header).len = len;
-            (*header).data = &mut (*(*header).shared.as_mut_ptr()).buffer_start;
-            (*(*header).shared.as_mut_ptr()).count = RefCount::new(1);
+            (*header).count = RefCount::new(1);
+            (*header).data = &mut (*header).buffer_start;
         }
         header
     }
 
     fn duplicate(&mut self) -> *mut Header {
         if self.flags & REFERENCE_FLAG == 0 {
             // If this is not a "fast pass" string then simply increment the reference count.
-            unsafe {
-                (*self.shared.as_ptr()).count.add_ref();
-                self
-            }
+            self.count.add_ref();
+            self
         } else {
             // Otherwise, allocate a new string and copy the value into the new string.
             let copy = Header::alloc(self.len);
+            // SAFETY: since we are duplicating the string it is safe to copy all data from self to the initialized `copy`.
             unsafe {
-                core::ptr::copy_nonoverlapping(self.data, (*copy).data, self.len as usize + 1);
+                core::ptr::copy_nonoverlapping(self.data, (*copy).data, self.len as usize);
```

Comment:
Except that `HSTRING` buffers are null terminated.

Reply:
Ok but then removing this +1 here is correct though, right?

Reply:
I was all turned around. The `alloc_size` (i.e., `let alloc_size = core::mem::size_of::<Header>() + 2 * len as usize`) does not take into account the terminating null because it's already accounted for by the extra `u16` allocated in the `Header`.

Reply:
With that commit, if you drop the +1 it will fail the test.